sasautos/h54s.sas

```diff
@@ -475,85 +475,47 @@ options NOQUOTELENMAX LRECL=32000 spool;
     %return;
   %end;
 
-
-  data _null_;
-    call symput('qc', '"');
-    call symput('pf', "%upcase(&dsn)");
-    call symput('dc', '$');
-    call symput('dt_', 'dt_');
+  proc contents noprint data=&libn..&dsn out=tempCols(keep=name type length);
   run;
 
-  proc sql;
-    create table tempCols as
-    select upcase(name) as name, type, length from dictionary.columns 
-    where upcase(memname)="%upcase(&dsn)" and libname="%upcase(&libn)";
-  quit;
-
-  %let totalCols = &sqlObs;
-
-  proc sql;
-    select trim(name), trim(type), length into :name1-:name999, :type1-:type999, :length1-:length999
-    from tempCols;
-  quit;
-
   * get first and last column names;
-
-  data tempCols;
+  data _null_;
     set tempCols end=lastcol;
+    name=upcase(name);
     if _n_ = 1 then do;
-      call symput('firstCol', strip(name));
+      call symputx('firstCol',name,'l');
     end;
+    call symputx(cats('name',_n_),name,'l');
+    call symputx(cats('type',_n_),type,'l');
+    /* char vars are urlencoded and lengthened to 30000 in next step */
+    if type=2 then call symputx(cats('length',_n_),30000,'l');
+    else call symputx(cats('length',_n_),length,'l');
     if lastcol then do;
-      call symput('lastCol', strip(name));
+      call symputx('lastCol',name,'l');
+      call symputx('totalCols',_n_,'l');
     end;
   run;
-
```

Comment:
use of datastep instead of proc sql to create macvars.
1 - not limited to 999
2 - applies conditional logic to character vars, setting them all to 30,000 length (this is basically all the second metadata step did, which has been removed)